packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -0,0 +1,407 @@
+  /// after the layout algorithm has finished and the non-visible children are
+  /// to be removed.
+  ///
+  /// The default implementation calls [remove], which removes the child from
+  /// the child list.
+  ///
+  /// The index of the given child can be obtained using the [indexOf] method,
+  /// which reads it from the [SliverMultiBoxAdaptorParentData.index] field of the child's
+  /// [parentData].
+  void removeChild(RenderBox child);
+
+  /// Called to estimate the total scrollable extents of this object.
+  ///
+  /// Must return the total distance from the start of the child with the
+  /// earliest possible index to the end of the child with the last possible
+  /// index.
+  double estimateScrollOffsetExtent({
+    int firstIndex,
+    int lastIndex,
+    double leadingScrollOffset,
+    double trailingScrollOffset,
+  });
+}
+
+class SliverMultiBoxAdaptorParentData extends SliverLogicalParentData with ContainerParentDataMixin<RenderBox> {
+  int index;
+
+  @override
+  String toString() => 'index=$index; ${super.toString()}';
+}
+
+// ///
+// ///
+// /// The contract for adding and removing children from this render object is
+// /// more strict than for normal render objects:
+// ///
+// /// - Children can be removed except during a layout pass if they have already
+// ///   been laid out during that layout pass.
+// /// - Children cannot be added except during a call to [allowAdditionsFor], and
+// ///   then only if there is no child correspending to that index (or the child
+// ///   child corresponding to that index was first removed).
+// ///
+// /// ## Writing a RenderSliverBlock subclass
+// ///
+// /// There are three methods to override:
+// ///
+// /// - [createChild], which must create the child for the given index, and then
+// ///   insert it in the given location in the child list.
+// ///
+// /// - [removeChild], which is called when the given child is cleaned up,
+// ///   and which should remove the given child from the child list.
+// ///
+// /// - [estimateScrollOffsetExtent], which should return the total extent (e.g.
+// ///   the height, if this is a vertical block) of all the children.
+abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
+  with ContainerRenderObjectMixin<RenderBox, SliverMultiBoxAdaptorParentData>,
+       RenderSliverHelpers {
+
+  RenderSliverMultiBoxAdaptor({
+    @required RenderSliverBoxChildManager childManager
+  }) : _childManager = childManager {
+    assert(childManager != null);
+  }
+
+  @override
+  void setupParentData(RenderObject child) {
+    if (child.parentData is! SliverMultiBoxAdaptorParentData)
+      child.parentData = new SliverMultiBoxAdaptorParentData();
+  }
+
+  @protected
+  RenderSliverBoxChildManager get childManager => _childManager;
+  final RenderSliverBoxChildManager _childManager;
+
+  int _currentlyUpdatingChildIndex;
+
+  void allowAdditionsFor(int index, VoidCallback callback) {
```

Comment:
I call this from the element.  Before there was a private trampoline to call it.